File: internal/replay/hash.go
// Package replay re-executes a recorded run against its own recording:
// cached model responses, recorded tool outputs, and an edited value at the
// span the user chose.
package replay

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"strings"
)

// The hashes below are recomputed by the SDK runner from live call arguments,
// so both sides must serialize identically: arrays keep order, HTML escaping
// is off, and nothing is reparsed. capybara/_replay.py mirrors this file.

// HashLLMRequest identifies a model call by its model and message texts.
func HashLLMRequest(model string, messages []Message) string {
	canonical := make([]any, 0, 2)
	pairs := make([][2]string, 0, len(messages))
	for _, m := range messages {
		pairs = append(pairs, [2]string{m.Role, MessageText(m.Body)})
	}
	canonical = append(canonical, model, pairs)
	return sum(marshal(canonical))
}

// HashToolCall identifies a tool call by name and its recorded argument text,
// which the SDK serialized and the runner reproduces verbatim.
func HashToolCall(tool, arguments string) string {
	return sum([]byte(tool + "\x00" + arguments))
}

// Message is one recorded conversation turn.
type Message struct {
	Role string
	Body string
}

// MessageText reduces a recorded message body to the plain text a live call
// would carry: instrumentors serialize parts as JSON, plain prompts as text.
func MessageText(body string) string {
	var value any
	dec := json.NewDecoder(strings.NewReader(body))
	dec.UseNumber()
	if err := dec.Decode(&value); err != nil {
		return strings.TrimSpace(body)
	}
	return strings.TrimSpace(partsText(value))
}

func partsText(value any) string {
	switch v := value.(type) {
	case string:
		return v
	case []any:
		var b strings.Builder
		for _, el := range v {
			b.WriteString(partsText(el))
		}
		return b.String()
	case map[string]any:
		for _, key := range []string{"content", "text"} {
			if inner, ok := v[key]; ok {
				return partsText(inner)
			}
		}
	}
	return ""
}

func marshal(v any) []byte {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(v); err != nil {
		return nil
	}
	return bytes.TrimRight(buf.Bytes(), "\n")
}

func sum(b []byte) string {
	h := sha256.Sum256(b)
	return hex.EncodeToString(h[:])
}

File: sdk/src/capybara/_hash.py
"""Request hashing, mirrored by internal/replay/hash.go; both must agree."""

from __future__ import annotations

import hashlib
import json
from typing import Any


def hash_llm_request(model: str, messages: list[tuple[str, str]]) -> str:
    payload = [model, [[role, text] for role, text in messages]]
    return _sum(json.dumps(payload, separators=(",", ":"), ensure_ascii=False))


def hash_tool_call(tool: str, arguments: str) -> str:
    return _sum(tool + "\x00" + arguments)


def message_text(body: str) -> str:
    """Reduce a recorded message body to the plain text a live call carries."""
    try:
        value = json.loads(body)
    except ValueError:
        return body.strip()
    return _text(value).strip()


def value_text(value: Any) -> str:
    """Reduce a live message content, which may be blocks, to plain text."""
    if isinstance(value, str):
        return value.strip()
    return _text(value).strip()


def _text(value: Any) -> str:
    if isinstance(value, str):
        return value
    if isinstance(value, list):
        return "".join(_text(element) for element in value)
    if isinstance(value, dict):
        for key in ("content", "text"):
            if key in value:
                return _text(value[key])
    return ""


def _sum(payload: str) -> str:
    return hashlib.sha256(payload.encode("utf8")).hexdigest()
